### ISSUE TITLE
SigmoidCrossEntropyLossLayer::Forward_gpu() is not implemented

### ISSUE THREAD

**Issue by liuyipei** [BVLC/caffe#3004]
I just wanted to point out that there are probably many people out there who are interested in having a forward GPU computation for SigmoidCrossEntropyLossLayer on GPU. Right now, so much time is spent shuffling data back and forth, that much of the improvement from using GPU at all is lost.

I would love to do this myself, except that I know very little: things like gpu memory allocations/deallocations, gpu loops, or not knowing diff_ responsibilities trouble me. 

A PR addressing this would be an obvious improvement to caffe.

@slayton58
Obviously your time is very valuable, but could you take a look when a chance arises?

**Comment by liuyipei** [comment]
Here is a bit more information. 
The formula used in the cpu forward pass is somewhat cryptic and maybe not the most efficient possible.

Let p be the target (in {0,1} or [0,1]).
Let q be the activation to be passed in. (any real)

The computation needed is:
"p log (sigmoid(q)) + (1-p) log(1-sigmoid(q))"  

This mathematically reduces to:
"-q(1-p) -log(1 + exp(-q))"


The current code in the CPU forward pass reads as something which is numerically equivalent, but mildly obfucated, depending on one's perspective.
"q * (p - (q >= 0)) - log(1 + exp(q - 2 * q * (q >= 0)))"

Obviously, I don't actually know the original intent, so maybe there was something magically fast about the way the CPU expression was written. To me it seems like it was an attempt to "bake in" the observation of symmetry in L(p, q) = L(1-p, -q). Where L denotes the loss term as a function of p and q.

(This is summarized from this caffe users post:

https://groups.google.com/forum/#!topic/caffe-users/6IwLKO2rtDE
)


Here are some python code (depends on numpy) one can copy and paste to verify the above expressions numerically for convenience.

```python
import numpy as np
input_data = [-0.3, -0.1, 0.1, 0.1]
target =[.5, 0, 1, 0]
sigmoid = lambda x: 1.0/(1.0+np.exp(-x))


for i in range(len(input_data)):
  loss = input_data[i] * (target[i] - (input_data[i] >= 0)) - np.log(1 + np.exp(input_data[i] - 2 * input_data[i] * (input_data[i] >= 0)))
  literal_loss = target[i] * np.log(sigmoid(input_data[i])) + (1-target[i]) *np.log(1-sigmoid(input_data[i]))
  my_loss = -input_data[i] * (1- target[i]) - np.log(1 + np.exp(-input_data[i]))
  loss2 = input_data[i] * (target[i] - (input_data[i] >= 0)) - np.log(1 + np.exp(-np.abs(input_data[i])))
  print loss, literal_loss, my_loss, loss2

```

**Comment by darrengarvey** [comment]
I have had a go at two things:

1) Changing (simplifying) the CPU implementation as per your comments above. This appears to give a ~5% performance improvement without affecting the results. I'm relying on your python test as I'm not sure I follow what you mean about the "baking in" of the symmetry of L(p, q) = L(1-q, -q). Could you please provide a reference to help me understand why this is the case?

2) Implement the Forward_gpu() method. I have failed to correctly use caffe_gpu_asum() as I imagine would be a good thing to use. It is ~10% faster but giving a different result to what I *think* should be an equivalent CPU implementation. Still, what I've done (add the vectorised loss that's calculated on the GPU in a loop) also isn't equivalent to caffe_cpu_asum(), which gives the same "aberration" as caffe_gpu_asum().

The GPU implementation of Forward_gpu() makes training much faster (~7x) and gives the same results on the test I was running.

I pushed a branch to:

https://github.com/darrengarvey/caffe/tree/SigmoidCrossEntropyLossLayer-forward-gpu-implementation

Do you want to have a look and see if it works for you?

**Comment by liuyipei** [comment]
> I'm not sure I follow what you mean about the "baking in" of the symmetry of L(p, q) = L(1-q, -q). Could you please provide a reference to help me understand why this is relevant?

This is just me speculating on what the original implementation was trying to "say". i.e. I am trying to contrive a perspective from which the earlier implemented equations/code is natural to read. It might have been written with the symmetry of the sigmoid in mind. This is not particularly relevant and just a point of potential interest for me.

**Comment by liuyipei** [comment]
Thank you so much for this great work @darrengarvey

**Comment by liuyipei** [comment]
@darrengarvey
The for stopping criterion from the original cpu forward pass actually looks tentatively like a bug. Shouldn't it be i < num?

From what I understand, num is the mini-batch batch size, whereas count refers to the product of all 4 dimension sizes.
```
  for (int i = 0; i < count; ++i) {
    loss -= input_data[i] * (target[i] - (input_data[i] >= 0)) -
        log(1 + exp(input_data[i] - 2 * input_data[i] * (input_data[i] >= 0)));
  }
  top[0]->mutable_cpu_data()[0] = loss / num;
```

Here is a relevant snippet from blob.cpp and blob.hpp respectively
```
template <typename Dtype>
void Blob<Dtype>::Reshape(const vector<int>& shape) {
  CHECK_LE(shape.size(), kMaxBlobAxes);
  count_ = 1;
  shape_.resize(shape.size());
  for (int i = 0; i < shape.size(); ++i) {
    CHECK_GE(shape[i], 0);
    CHECK_LE(shape[i], INT_MAX / count_) << "blob size exceeds INT_MAX";
    count_ *= shape[i];
    shape_[i] = shape[i];
  }
  if (count_ > capacity_) {
    capacity_ = count_;
    data_.reset(new SyncedMemory(capacity_ * sizeof(Dtype)));
    diff_.reset(new SyncedMemory(capacity_ * sizeof(Dtype)));
  }
}
```

```
  inline int count() const { return count_; }
```

**Comment by liuyipei** [comment]
I am really grasping at straws here. Base on looking at this snippet from test_math_functions.cpp, what if you explicitly added the template type to the caffe_gpu_asum to the call the way they do?

```
TYPED_TEST(GPUMathFunctionsTest, TestAsum) {
  int n = this->blob_bottom_->count();
  const TypeParam* x = this->blob_bottom_->cpu_data();
  TypeParam std_asum = 0;
  for (int i = 0; i < n; ++i) {
    std_asum += std::fabs(x[i]);
  }
  TypeParam gpu_asum;
  caffe_gpu_asum<TypeParam>(n, this->blob_bottom_->gpu_data(), &gpu_asum);
  EXPECT_LT((gpu_asum - std_asum) / std_asum, 1e-2);
}
```

**Comment by darrengarvey** [comment]
> The for stopping criterion from the original cpu forward pass actually looks tentatively like a bug. Shouldn't it be i < num?

You may well be correct. In the example you had given me the losses I was seeing were huge (~516). Also the loss itself it divided by `num` at the end rather than `count` as you might expect if `count` was the correct value to be summing over.

I need to dig about a bit more about this loss layer to be sure but one nice thing is that now I have a loss of `0.692655` for the on-CPU loop and `0.692656` when using `caffe_gpu_asum()`, which is more like it. Of course since `num` is < `count` this is also even faster!

@liuyipei - Do you have some more information on where I might look to find a way of constructing a more simplified test that can give me an expected loss that can be verified against? I'll dig myself but it seems you're more familiar with the theory of sigmoid cross-entropy loss and I'd like to do more testing before raising a PR for these changes.

**Comment by liuyipei** [comment]
@darrengarvey Sadly, after more thought, I ended up retracting my comment about the for's end criterion (which you quoted and I deleted).  The layer needs to compute a loss term for every spatial position, for every channel, and for every mini-batch member, then sums it up. The particular definition they wrote in their docs specifies `count` (number of mini-batches) for the denominator: see 
http://caffe.berkeleyvision.org/doxygen/classcaffe_1_1SigmoidCrossEntropyLossLayer.html

That said, the fact that the your two values now match up is encouraging because it gives us clues about what is going on.

Here is a python snippet for computing the loss values for the first member of the first batch for mnist_train using that config I gave you over email earlier (where a single MNIST data layer connects directly to SigmoidCrossEntropyLoss's two bottoms).

``` python
import numpy as np
import lmdb
import matplotlib.pyplot as plt

caffe_home = '/home/yiliu/caffe/caffe'
env = lmdb.open(caffe_home + '/examples/mnist/mnist_train_lmdb', readonly=True)
with env.begin() as txn:
    raw_datum = txn.get(b'00000000')

datum = caffe.proto.caffe_pb2.Datum()
datum.ParseFromString(raw_datum)

x = np.fromstring(datum.data, dtype=np.uint8).reshape(datum.channels, datum.height, datum.width)
xx = x[0,:,:] 
#plt.imshow(xx)

z = xx / 256
sigmoid_cross_entropy_terms = - z * (1- z) - np.log(1 + np.exp(-z))
```

**Comment by liuyipei** [comment]
I finally actually got around to running your code.  Sorry for the delays! The output from using your caffe_gpu_asum appears fine. Both in terms of visual sanity (i.e. running the config I gave you and looking at the loss numbers), and in terms of running the existing sigmoid_cross_entropy_loss_layer.testbin.

**Comment by paulfitz** [comment]
Just to note for others trying to follow this: there is a later post on the caffe users thread mentioned in https://github.com/BVLC/caffe/issues/3004#issuecomment-136883394 pointing out that the cpu implementation goes out of its way to avoid taking the exponential of a positive number:

https://groups.google.com/d/msg/caffe-users/6IwLKO2rtDE/aMl1cvoPDQAJ
